pkg/util/fxutil/logging/tracer.go

```diff
@@ -0,0 +1,146 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package logging
+
+import (
+	"os"
+	"sync"
+	"time"
+
+	"go.uber.org/fx/fxevent"
+)
+
+const (
+	serviceName = "datadog-agent-fx"
+)
+
+type traceSpanType int
+
+const (
+	fxConstructorType traceSpanType = iota
+	fxOnStartHookType
+)
+
+// TraceSpan represents a captured span from Fx events.
+type TraceSpan struct {
+	Type     traceSpanType
+	Resource string
+	Start    int64 // Unix timestamp in nanoseconds
+	Duration int64 // Duration in nanoseconds
+	Error    error
+}
+
+// fxTracingLogger implements fxevent.fxTracingLogger interface to capture Fx lifecycle events
+// and send them as traces to Datadog.
+type fxTracingLogger struct {
+	fxLogger       fxevent.Logger
+	agentLogger    Logger
+	flavor         string
+	mu             sync.Mutex
+	spans          []*TraceSpan // Buffer spans during startup
+	startTime      time.Time    // Fx startup time
+	lifecycleStart time.Time    // Fx lifecycle start time
+	endTime        time.Time    // Fx completion time
+	err            error
+}
+
+// withFxTracer wraps the fxlogger with a fxTracingLogger.
+// Tracing is enabled when DD_FX_TRACING_ENABLED is set to true.
+// When enabled, it instruments Fx lifecycle events including component construction
+// and OnStart hooks, sending traces to the configured trace agent.
+func withFxTracer(fxlogger fxevent.Logger, agentLogger Logger, flavor string) fxevent.Logger {
+	if os.Getenv("DD_FX_TRACING_ENABLED") == "true" {
+		agentLogger.Infof("[Fx Tracing] Initialized - will capture component construction and OnStart hooks")
+		return &fxTracingLogger{
+			fxLogger:    fxlogger,
+			agentLogger: agentLogger,
+			flavor:      flavor,
+			spans:       make([]*TraceSpan, 0),
+		}
+	}
+
+	return fxlogger
+}
+
+// LogEvent implements the fxevent.Logger interface.
+func (l *fxTracingLogger) LogEvent(event fxevent.Event) {
+	notificationTime := time.Now()
+	l.mu.Lock()
+
+	switch e := event.(type) {
+	case *fxevent.LoggerInitialized:
+		l.handleLoggerInitialized(e, notificationTime)
+
+	case *fxevent.Run:
+		l.handleRun(e, notificationTime)
+
+	case *fxevent.OnStartExecuted:
+		l.handleOnStartExecuted(e, notificationTime)
+
+	case *fxevent.Started:
+		l.handleStarted(e, notificationTime)
+	}
+	l.mu.Unlock()
```

Comment:
Pretty subjective but we could lock only when needed
In particular just when adding a span to the list or updating a field